gordo_components/builder/azure_utils.py

```diff
@@ -0,0 +1,192 @@
+    run_id: str
+        Unique ID of MLflow run to log to.
+    """
+    experiment = client.get_experiment_by_name(experiment_name)
+
+    experiment_id = (
+        getattr(experiment, "experiment_id")
+        if experiment
+        else client.create_experiment(experiment_name)
+    )
+
+    return client.create_run(experiment_id, tags={"model_key": model_key}).info.run_id
+
+
+def get_loggables(
+    d: dict, parent_key: str = None
+) -> Iterator[Tuple[str, Union[str, int, float]]]:
+    """
+    Recursively generate key, value pairs form multilevel dictionary for logging
+
+    Parameters
+    ----------
+    d: dict
+        Dictionary to flatted to loggable (key, value) pairs.
+    parent_key: str
+        Key of parent dict to pre-prend to loggable key.
+
+    Example
+    -------
+    >>> d = {"a1":2, "b1":{"c2":"text", "d2": 5.398, "e2":{"f3":2, "g3":"word"}}}
+    >>> list(get_loggables(d))
+    [('a1', 2),
+     ('b1_c2', 'text'),
+     ('b1_d2', 5.398),
+     ('b1_e2_f3', 2),
+     ('b1_e2_g3', 'word')]
+    """
+    for k, v in d.items():
+        if type(k) is not str:
+            raise ValueError("Key {k} must be of type 'str', not {type(k)}.")
+
+        key = f"{parent_key}_{k}" if parent_key else k
+        if type(v) is dict:
+            yield from get_loggables(v, parent_key=key)
+        else:
+            yield (key, v)
+
+
+def get_batch_kwargs(d: dict, timestamp: datetime = None, step: int = 0) -> dict:
+    """
+    Create flat lists of MLflow logging entities from multilevel dictionary
+
+    For more information, see the mlflow docs:
+    https://www.mlflow.org/docs/latest/python_api/mlflow.tracking.html#mlflow.tracking.MlflowClient.log_batch
+
+    Parameters
+    ----------
+    d: dict
+        Multi-level dictionary with (key, value) loggables to parse.
+    timestamp: datetime
+        Timezone aware timestamp of time to associate with all metrics logged.
+    step: int
+        Training step to associate with all metrics.
+
+    Returns
+    -------
+    batch_kwargs: dict
+        Dict with `metrics` and `params` lists for passing to
+        `mlflow.tracking.MlflowClient.log_batch`. The metrics key containes a list of
+        Mlflow `Metric` instances, and the params key contains a list of `Param`
+        instances.
+    """
+
+    def _is_number(s: Any):
+        """
+        Check that the argument can be cast to a number type
+        """
+        valid = True
+        try:
+            float(s)
+        except (TypeError, ValueError):
+            valid = False
+        return valid
+
+    if not timestamp:
+        timestamp = datetime.now(tz=UTC)
+    elif not timestamp.tzinfo:
+        raise ValueError(
+            f"Timestamp ({timestamp} needs to include timezone information"
+        )
+
+    if type(step) is not int:
+        raise ValueError(f"`step` argument must be of type '{str(int)}' not '{type(step)}'")
+
+    metrics = list()
+    params = list()
+    for k, v in get_loggables(d):
```

Comment:
Suppose this kind of makes sense, although I think it will be a bit random in that things that are "numeric" parameters will be logged as metrics. But not sure it really matters.

Reply:
I'm not sure how I would do it differently while still having it be able to generally take a dict of unkown structure. The difference between `Param` and `Metric` instances is the latter has a timestamp/step associated with them and they are populated at plottables on the frontend.